### PR TITLE
UI,CI,cmake,plugins: Add Caffeine.tv support

### PR DESCRIPTION
**Pull request by cogwheel**
### Description

Adds custom sign-in and an output for broadcasting to Caffeine.tv.

#### Sign-in

A custom UI was added for caffeine sign-in (see motivations below). Here is what the flow looks like:

before sign-in
signing in
after sign-in

Once signed in, a Caffeine dock window appears for controlling and viewing the broadcast. We have plans to add chat to this window in the future.

caffeine window

#### Caffeine output

The caffeine output is built around [libcaffeine](https://github.com/caffeinetv/libcaffeine), which handles authentication, broadcast setup/teardown, and takes the raw (unencoded) A/V from OBS to pass along to WebRTC.

The output also monitors for broadcast title changes from the caffeine window, and detects Caffeine-supported games coming to the foreground for categorizing the broadcast on our site.

#### Ancillary changes

In Settings > Stream, when you select a new service, the stream key field is cleared. If you select the original service without closing the window, the original stream key will be restored.

In the autoconfig wizard, the service panel now appears before the output settings.

### Motivation and Context

The general motivation is that Caffeine users want to use OBS Studio to broadcast on Caffeine, and we would prefer that it be available in the default installation rather than requiring people to use a separate fork.

Some of the motivations for specific changes are:

- Caffeine does not use stream keys or have an OAuth login page, hence:
    - Custom sign-in UI instead of browser-based
    - Ability to require custom sign-in rather than stream key
- Caffeine has strict bandwidth, resolution, and FPS limits and the WebRTC library handles encoding itself, hence:
    - using the raw output instead of the encoded output
    - disabling hardware encoder testing during autoconfig when Caffeine is selected
    - Reordering the autoconfig wizard so service is selected before settings
- The WebRTC library is enormous and time-consuming to build, difficult to link to from anything other than clang-cl on Windows, and we use some customizations on the stock version, hence wrapping WebRTC in our own libcaffeine layer rather than implementing a general WebRTC output
- To simplify implementation, the caffeine auth uses the stream key field under the hood to store the caffeine refresh token. Hence clearing the stream key field when switching services.
- We would like to have Caffeine supported in the default build configuration of OBS Studio, hence looking for libcaffeine in DepsPath.

### How Has This Been Tested?

- Functional tests by our QA department to verify that streaming (both to Caffeine and via RTMP), game capture, and other sources continue to work. Most tests were performed on latest versions of Windows 10 (64- and 32-bit) and macOS, with occasional tests on Windows 7 and 8.
- In the wild: we released this build to our users last week and are monitoring for issues.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [X] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [X] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [X] My code is not on the master branch.
- [X] The code has been tested.
- [ ] All commit messages are properly formatted and commits squashed where appropriate.

(Commits will be re-squashed onto the latest master, pending review)

This change is [Reviewable](https://reviewable.io/reviews/obsproject/obs-studio/1950)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -166,6 +166,7 @@ if(NOT INSTALLER_RUN)
 		endif()
 
 		find_package(Qt5Widgets ${FIND_MODE})
+		find_package(libcaffeine)
```
Comment (kkartaltepe) [inline CMakeLists.txt]: Does it need to be found both here and in UI?

File: UI/CMakeLists.txt [hunk 2]
```diff
@@ -39,6 +39,13 @@ else()
 	set(MIXER_ENABLED TRUE)
 endif()
 
+find_package(libcaffeine)
+if(LIBCAFFEINE_FOUND)
```
Comment (kkartaltepe) [inline UI/CMakeLists.txt]: will this overwrite CAFFEINE_ENABLED being set manually?

Comment (cogwheel) [reply UI/CMakeLists.txt]: Hmm it's not implemented as an option so that wasn't considered. I'll make that change

**Comment by pas256** [comment]
Looks like this needs a rebase @cogwheel to resolve the conflicts

**Comment by cogwheel** [comment]
> 
> Looks like this needs a rebase @cogwheel to resolve the conflicts

I'm going to do another rebase after implementing review changes

File: CMakeLists.txt (hunk 1 above)
Comment (cogwheel) [reply CMakeLists.txt]: I removed the redundant ones

File: UI/CMakeLists.txt (hunk 2 above)
Comment (cogwheel) [reply UI/CMakeLists.txt]: I added a BUILD_CAFFEINE option

Comment (kkartaltepe) [reply UI/CMakeLists.txt]: Cool looks good.

**Comment by cogwheel** [comment]
Squashed and rebased onto master

**Comment by WizardCM** [comment]
Friendly bump: there are some merge conflicts. Please also squash any formatting & merge commits

**Comment by Xaymar** [comment]
Will pick this up tomorrow, we're currently rebasing our codebase onto 24.x so this shouldn't take too long.

**Comment by Xaymar** [comment]
I regret saying that it shouldn't take too long. There's still issues with the UI part that make no sense, so it's still not entirely finished yet. But it should compile with current master.

Edit: This would greatly benefit from a full rework of the outputs design, but that is still very far off - if it even gets made at all (lack of input is currently the problem).

**Comment by Xaymar** [comment]
Updated to `master` once again. Internal testing is still ongoing.

**Comment by cameronelliott** [comment]
This is quite interesting, although if it hasn't been said already, it would seem the different efforts to add WebRTC support to OBS should maybe be talking a little bit, so there are not multiple disparate efforts to add WebRTC. I would think the evolving WHIP/WISH protocol would be a good candidate for ingress. WISH is proceeding down the IETF path to be come an IETF standard. https://datatracker.ietf.org/wg/wish/about/

**Comment by dodgepong** [comment]
More details on our thoughts on WebRTC can be found in the adjacent FTL discussion here: https://github.com/obsproject/obs-studio/discussions/4021

Honestly, I'm surprised this PR is still open -- I thought it was closed, and it probably ought to be pending more foundational integration of WebRTC support in OBS. I'm closing it now. This can be revisited once OBS actually supports WebRTC output in a more generally-applicable fashion, and we'd be happy to talk to Caffeine about how we might want to see such an integration when that time comes.

**Comment by cameronelliott** [comment]
@dodgepong Thanks for the link, I'll be reading that long discussion carefully.


I'm working on a yet unreleased SFU that takes WHIP (precursor to WISH)

It works great with older Cosmo releases of their OBS+WHIP fork.

I'd like to be working with the best candidate for what will become the official OBS WISH incarnation.

Am I best-off using the Cosmo forks for testing & feedback on WISH support?

Or, is there a different branch or fork of OBS with alpha or beta work on WISH that I would be 
better off testing against and providing feedback on?

Sorry if this question is too early in the progress of the WebRTC effort, I just want to throw it out there.

Thanks,
Cameron

**Comment by dodgepong** [comment]
While I haven't examined the Cosmo fork closely, we currently have no plans to use any of its code in the official OBS implementation. We'd really like to avoid using libwebrtc if possible, and are investigating alternate libraries. That said, I will defer to @DDRBoxman on matters related to WebRTC support in OBS.

**Comment by Xaymar** [comment]
The last information I have about this is that Caffeine switched direction, and that was almost a year ago now.

**Comment by cameronelliott** [comment]
@DDRBoxman are you sharing whether a stack has been decided for WebRTC signalling at this time? Are you near a point where you need testers?  (sorry for this question)
I'm working heavily in Go regarding WebRTC, so can't really help much in C/C++ land on the options.
Here is what I do know of:
- there is of course the well-known libwebrtc codebase 
- there is the [stack Centricular did](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/tree/master/ext/webrtc) which has been adopted by gStreamer
- there is a [pure-C stack](https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c) by Sean D., author of Pion
- there [was a DTLS implementation](https://github.com/sustrik/libdill/commit/d4611fe7a7edcecd65b1e207fc631c6afb2ff33e) in Sustrick's nifty green-threads libdill project. But DTLS alone is a small part of WebRTC, obviously.

Anyway, if you want someone to bounce ideas about WebRTC off of, or need someone for testing an OBS alpha/beta, I'd like to help if I can.

**Comment by cameronelliott** [comment]
Hmm, one curious question comes, if you guys don't use the Cosmo codebase, I wonder whether you will plan to support the 3x level output possible with Simulcast (for example with Chrome and WebRTC simulcast you get 3 RTP SSRCs, and three video levels at roughly 2.5mbps, 1.5 mbps, and 140kbps)
Well just something I thought might be interesting to ask about.

**Pull request closed without merging**